### DISCUSSION TITLE
How to set default column type for all columns? I want to use X as the default value

### DISCUSSION THREAD

**Discussion by rodrigo-morales-1** [lvjr/tabularray#408] [Q&A]
# Table of contents

- [The context](#orgd28c67c)
- [The problem](#org6aebb4d)
- [A hacky workaround](#org5941edf)
- [The question](#org2c61a51)
- [Additional information](#orgac08412)

<a id="orgd28c67c"></a>

# The context

I've defined my own environment using `NewTblrEnviron`. I can use this environment regardless of the number of columns. Plase see minimal working example below.

```tex
\documentclass{article}

\usepackage{tabularray}
\usepackage{xcolor}

\begin{document}

\NewTblrEnviron{mytblr}
\SetTblrInner[mytblr]{
  row{odd} = {red!10},
  row{even} = {red!40}
}

The following is a table of two columns.

\begin{mytblr}{}
1 & foo \\
2 & foo foo \\
3 & foo foo foo \\
\end{mytblr}

The following is a table of three columns.

\begin{mytblr}{}
1 & foo & bar\\
2 & foo foo & bar bar \\
3 & foo foo foo & bar bar bar \\
\end{mytblr}

The following is a table of four columns.

\begin{mytblr}{}
1 & foo & bar & fizz \\
2 & foo foo & bar bar & fizz fizz \\
3 & foo foo foo & bar bar bar & fizz fizz fizz \\
\end{mytblr}

\end{document}
```

![image](https://github.com/lvjr/tabularray/assets/74389646/d6cd3665-72e1-4ac4-8aa2-9e77b6fc29ed)

I now want the type of all the column in all tables that use that environment to be `X`. I know I can do this by explicitly specifying `colspec = {<<as much X's as the number of columns>>}` in each table (see code block below).

```tex
\documentclass{article}

\usepackage{tabularray}
\usepackage{xcolor}

\begin{document}

\NewTblrEnviron{mytblr}
\SetTblrInner[mytblr]{
  row{odd} = {red!10},
  row{even} = {red!40},
}

The following is a table of two columns.

\begin{mytblr}{colspec=XX}
1 & foo \\
2 & foo foo \\
3 & foo foo foo \\
\end{mytblr}

The following is a table of three columns.

\begin{mytblr}{colspec=XXX}
1 & foo & bar\\
2 & foo foo & bar bar \\
3 & foo foo foo & bar bar bar \\
\end{mytblr}

The following is a table of four columns.

\begin{mytblr}{colspec=XXXX}
1 & foo & bar & fizz \\
2 & foo foo & bar bar & fizz fizz \\
3 & foo foo foo & bar bar bar & fizz fizz fizz \\
\end{mytblr}

\end{document}
```

![image](https://github.com/lvjr/tabularray/assets/74389646/f7f43838-014a-4c03-8070-6aa00c7ceb0f)

<a id="org6aebb4d"></a>

# The problem

Now, let's imagine that my document has a huge number of tables, I want the type of all columns in those tables to be `X` and I don't want to manually specify `colspec = {XXX}` in each table. How to set the type of all columns for all tables?

# The question

How to set the column type for all columns in tables that use an environment that I've previously defined using `NewTblrEnviron`.

<a id="org5941edf"></a>

# A hacky workaround

A dirty solution would be to set a long `colspec` and, thus, we make sure that all columns use that column type. However, I feel this is a hacky way of doing and there should be a more cleaner, simpler and readable way (probably, something like `colspec = X*\ncolumns`).

```tex
\documentclass{article}

\usepackage{tabularray}
\usepackage{xcolor}

\begin{document}

\NewTblrEnviron{mytblr}
\SetTblrInner[mytblr]{
  row{odd} = {red!10},
  row{even} = {red!40},
  colspec = XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
}

The following is a table of two columns.

\begin{mytblr}{}
1 & 10 \\
2 & 20 \\
3 & 30 \\
\end{mytblr}

The following is a table of three columns.

\begin{mytblr}{}
1 & 10 & 100 \\
2 & 20 & 200 \\
3 & 30 & 300 \\
\end{mytblr}

The following is a table of four columns.

\begin{mytblr}{}
1 & 10 & 100 & 1000 \\
2 & 20 & 200 & 2000 \\
3 & 30 & 300 & 3000 \\
\end{mytblr}

The following is a table of five columns.

\begin{mytblr}{}
1 & 10 & 100 & 1000 & 10000 \\
2 & 20 & 200 & 2000 & 20000 \\
3 & 30 & 300 & 3000 & 30000 \\
\end{mytblr}

\end{document}
```

![image](https://github.com/lvjr/tabularray/assets/74389646/d93fa4fe-a34f-4ccb-9565-e6ca237a44fb)

<a id="org2c61a51"></a>

<a id="orgac08412"></a>

# Additional information

Someone asked [a similar question](https://tex.stackexchange.com/questions/686262/change-default-column-type-in-tabularray) in TeX Stack Exchange.

**Comment by muzimuzhi** [comment]
You can use option `columns={<column settings>}` to set style for all columns. It's currently (`tabularray` 2023A (2023-03-01)) documented in sec. 2.5.1 "Rows and Columns in New Interfaces".

Hence `columns={co=1}` will set all columns to `X` columns, which is defined as `Q[co=1,#1]`, see package doc sec. 2.6.2 "Column Types".

```tex
\documentclass{article}

\usepackage{tabularray}
\usepackage{xcolor}

\begin{document}

\NewTblrEnviron{mytblr}
\SetTblrInner[mytblr]{
  row{odd} = {red!10},
  row{even} = {red!40},
  columns = {co=1}
}

The following is a table of two columns.

\begin{mytblr}{}
1 & 10 \\
2 & 20 \\
3 & 30 \\
\end{mytblr}

The following is a table of three columns.

\begin{mytblr}{}
1 & 10 & 100 \\
2 & 20 & 200 \\
3 & 30 & 300 \\
\end{mytblr}

The following is a table of four columns.

\begin{mytblr}{}
1 & 10 & 100 & 1000 \\
2 & 20 & 200 & 2000 \\
3 & 30 & 300 & 3000 \\
\end{mytblr}

The following is a table of five columns.

\begin{mytblr}{}
1 & 10 & 100 & 1000 & 10000 \\
2 & 20 & 200 & 2000 & 20000 \\
3 & 30 & 300 & 3000 & 30000 \\
\end{mytblr}

\end{document}
```
![image](https://github.com/lvjr/tabularray/assets/6376638/c2fc62ec-be9b-4923-a917-0b52467e7fa5)

**Comment by muzimuzhi** [comment]
Perhaps what you really wanted is something like `columns type=<single column type>` so one can use `columns type=X`.  Then `tabularray` has to check if `<single column type>` really contains only one column type.

**Reply by muzimuzhi** [reply]
<details><summary>An experimental implementation for new option <code>columns type=&lt;single column type&gt;</code></summary>
<p>

```tex
\documentclass{article}

\usepackage{tabularray}
\usepackage{xcolor}

\ExplSyntaxOn
\keys_define:nn { tblr }
  {
    columns~ type .code:n = \__tblr_set_every_column_to_type:n {#1}
  }
\clist_gput_right:Nn \g__tblr_table_known_keys_clist { columns~ type }

% similar to \__tblr_parse_colrow_spec:nn
\cs_new_protected:Npn \__tblr_set_every_column_to_type:n #1
  {
    \tl_gset:Nn \g__tblr_column_or_row_tl {column}
    \tl_gset:Nn \g__tblr_expanded_colrow_spec_tl {#1}
    \__tblr_expand_colrow_spec:N \g__tblr_expanded_colrow_spec_tl
    \group_begin:
    \exp_args:Nc \RenewDocumentCommand { tblr_primitive_column_type_ Q } { O{} }
      {
        % this is the command option `columns` will invoke
        \__tblr_set_every_column_aux:n { ##1 }
        % skip any remaining column type(s)
        \__tblr_clean_to_scan:w
      }
    \__tblr_execute_colrow_spec:N \g__tblr_expanded_colrow_spec_tl
    \group_end:
  }

\cs_new:Npn \__tblr_clean_to_scan:w #1 \scan_stop: {}
\ExplSyntaxOff

\begin{document}

\NewTblrEnviron{mytblr}
\SetTblrInner[mytblr]{
  row{odd} = {red!10},
  row{even} = {red!40},
  columns type = {X}
}

The following is a table of two columns.

\begin{mytblr}{}
1 & 10 \\
2 & 20 \\
3 & 30 \\
\end{mytblr}

The following is a table of three columns.

\begin{mytblr}{}
1 & 10 & 100 \\
2 & 20 & 200 \\
3 & 30 & 300 \\
\end{mytblr}

The following is a table of four columns.

\begin{mytblr}{}
1 & 10 & 100 & 1000 \\
2 & 20 & 200 & 2000 \\
3 & 30 & 300 & 3000 \\
\end{mytblr}

The following is a table of five columns.

\begin{mytblr}{}
1 & 10 & 100 & 1000 & 10000 \\
2 & 20 & 200 & 2000 & 20000 \\
3 & 30 & 300 & 3000 & 30000 \\
\end{mytblr}

\end{document}
```

![image](https://github.com/lvjr/tabularray/assets/6376638/436a615d-f498-4c8f-b364-1f2cf1434e65)

</p>
</details>
